lib/src/tag_detection.dart

```diff
@@ -824,125 +830,154 @@ class Tagger {
+          final prunedLibraryGraph = trustDeclarations
+              ? LibraryGraph(
+                  _session, flutterPlatform.runtime.declaredVariables,
+                  isLeaf: declaredPlatformDetector._isFlutterPlugin)
+              : libraryGraph;
+
+          final prunedViolationFinder = _PlatformViolationFinder(
+              flutterPlatform,
+              prunedLibraryGraph,
+              declaredPlatformDetector,
+              _pubspecCache,
+              runtimeViolationFinder(
+                  prunedLibraryGraph,
+                  flutterPlatform.runtime,
+                  (List<Uri> path) => Explanation(
+                      'Package not compatible with runtime ${flutterPlatform.runtime.name} of ${flutterPlatform.name}',
+                      'Because of ${LibraryGraph.formatPath(path)}')));
+          // Report only the first non-pruned violation as Explanation
+          final firstNonPrunedViolation = _topLibraries
+              .map(violationFinder._findPlatformViolation)
+              .firstWhere((e) => e != null, orElse: () => null);
+          if (firstNonPrunedViolation != null) {
+            explanations.add(firstNonPrunedViolation);
+          }
 
-        // Tag is supported, if there is no pruned violations
-        final supports = _topLibraries
-            .map(prunedViolationFinder._findPlatformViolation)
-            .every((e) => e == null);
+          // Tag is supported, if there is no pruned violations
+          final supports = _topLibraries
+              .map(prunedViolationFinder._findPlatformViolation)
+              .every((e) => e == null);
 
-        if (supports) {
-          tags.add(flutterPlatform.tag);
+          if (supports) {
+            tags.add(flutterPlatform.tag);
+          }
         }
       }
+    } on _TagException catch (e) {
+      explanations.add(Explanation('Tag detection failed.', e.message));
+      return;
     }
   }
 
   /// Adds tags for the Dart runtimes that this package supports to [tags].
   ///
   /// Adds [Explanation]s to [explanations] for runtimes not supported.
   void runtimeTags(List<String> tags, List<Explanation> explanations) {
-    if (_isBinaryOnly) {
-      tags.addAll(<String>[Runtime.nativeAot.name, Runtime.nativeJit.name]);
-    } else if (_topLibraries.isEmpty) {
-      Explanation('No top-level libraries found',
-          'Cannot assign runtime tags, because no .dart files where found in lib/');
-    } else {
-      final dartSdkViolationFinder = _SdkViolationFinder(
-          _packageGraph, _Sdk.dart, _pubspecCache, _session);
-      if (dartSdkViolationFinder.findSdkViolation(packageName, _topLibraries) !=
-          null) {
-        // This is reported elsewhere
+    try {
+      if (_isBinaryOnly) {
+        tags.addAll(<String>[Runtime.nativeAot.name, Runtime.nativeJit.name]);
+      } else if (_topLibraries.isEmpty) {
+        Explanation('No top-level libraries found',
```

Comment:
shouldn't this be `explanations.add(...)`?

Reply:
yes!!